docs/DesignDocs/DebuggerSupportForMultiModuleEnclaves.md

```diff
@@ -0,0 +1,268 @@
+Debugger Support for Multi-Module Enclaves
+====
+
+This document describes extensions to the
+[Debugger Contract](/include/openenclave/internal/debugrt/host.h) to support
+debugging enclaves comprised of multiple modules.
+
+Introduction
+------------
+
+Currently, enclaves are built as a single module - a special elf executable that
+is loaded by OE SDK's enclave loader. Since enclaves are loaded by a special
+loader, unlike normal executables and shared libraries that are loaded by the
+platform's loader (e.g /lib64/ld-linux-x86-64.so.2), the debugger does not
+natively know how to load symbols for enclaves. Therefore, the host side OE SDK
+runtime runtime invokes following functions to enable the debugger to locate
+and debug an enclave.
+
+- `oe_result_t oe_debug_notify_enclave_created(oe_debug_enclave_t* enclave)`
+
+  This function is called immediately after an enclave has been loaded into
+  memory, but before it has been initialized/begun execution. The debugger uses
+  the following fields in `oe_debug_enclave_t` structure to load the symbols
+  and mark a region of program memory as being take up by the enclave:
+
+ ```c
+
+    typedef struct _debug_enclave_t
+	{
+		const char* path;
+		uint64_t path_length;
+
+		const void* base_address;
+		uint64_t size;
+	} oe_debug_enclave_t;
+	
+  ```
+
+  The `path` and `path_length` fields tell the debugger the path to the
+  enclave's ELF binary. The elf binary is expected to contain the debugging
+  information, if any, for the enclave. `oegdb` currently does not support
+  debugging enclaves whose debugging information has been stripped and moved to
+  a separate .dbg file.
+
+  The `base-address` field tells the debugger the memory address at which the
+  enclave has been loaded. The `size` field is not that important since the
+  debugger is interested only in the size of the `.text` section which it can
+  obtain from the ELF binary. The debugger associates the memory region
+  [base-address, base-address+size) with the enclave.
+
+- `oe_result_t oe_debug_notify_enclave_terminated(oe_debug_enclave_t* enclave)`
+
+  This function is called immediately after an enclave has been terminated and
+  it's memory has been deallocated. The debugger unloads the symbols for the enclave
+  and disassociates the memory region from the enclave.
+
+The debugger gains control and responds to the above notifications in the
+following ways:
+
+- Linux
+
+  The debugger sets hidden breakpoints in the above notification functions. When
+  they are called by the SDK's host-side runtime, the breakpoints are hit and
+  the debugger gets control. The debugger then reads the `oe_debug_enclave_t`
+  parameter to handle the notification.
+
+- Windows
+ 
+  Use of breakpoints by the debugger is not recommended. Instead, the Windows
```

Comment:
can you say why it's not recommended?  (Recommended by whom?)

Reply:
It's a windows symbol issue. Since PDBs are almost always separate from the DLL it's fairly easy to get into situations where you do not have the symbols available. That makes relying on the debugger setting breakpoints (via symbol data) a lot more fragile. It's much simpler on the debugger side to just receive an event by an exception as that will always work, regardless of whether or not symbols for the oe runtime dlll are available.

Reply:
@auott But enclaves are ELFs not DLLs so why are PDBs relevant for multi-module enclaves?

Reply:
It depends on where the load notification is coming from - not what is being loaded.
If the load notification comes from a DLL (similar to how the enclave load notification happens) then we have the PDB separation problem. If the notification is coming from inside an enclave, then we don't have the PDB problem.

My understanding was that both those scenarios are possible, though my previous response was about notifications coming from outside an enclave. For the latter case I don't see any issue with relying on symbols.

Reply:
One thing to keep in mind is that the contract supports Intel SDK's PE enclaves too. 
Instead of relying on debug symbols, would it be OK to rely on exported symbols that are guaranteed to be available even if debug symbols have been stripped off in a binary?
Then we could have both the ELF and PE enclaves export the functions and debugger ought to be able to put breakpoints on these even without debug symbols. However in that case, we'd also need the parameter to passed in a specific register so that the debugger can just read it without requiring debug symbols for the parameter.

Reply:
I think that would be work as well.
The register problem would likely need to be solved anyways as putting a breakpoint based on a symbol would run into the same problem as the breakpoint would likely be at the beginning of the prolog.